src/Access/IAccessStorage.h

```diff
@@ -31,6 +31,9 @@ struct AuthResult
     UUID user_id;
     /// Session settings received from authentication server (if any)
     SettingsChanges settings;
+
+    /// Indication whether user passed authentication.
+    bool authenticated;
```

Comment:
I don't really like this field because it makes things more complex. I mean it's simplier if `AuthResult` means only successful authentication; and if it's not successful an exception is thrown. This field is not necessary if `AccessControl::authenticate` did things in a bit different order (see [comment](https://github.com/ClickHouse/ClickHouse/pull/54737/files#r1466571255)).

Reply:
Thanks for the review,

What alternatives do we have?
The code requires `user_id` to find `EnabledQuota`, we can provide it using exception inheritance like that:

```cpp
class WrongPasswordException : public Exception
{
public:
WrongPasswordException(const PreformattedMessage & msg, UUID user_id_) 
  : Exception (msg, ErrorCodes::WRONG_PASSWORD)
  ,  user_id(user_id_)
{
}

const UUID user_id;
};
```
and throw it in `IAccessStorage::throwInvalidCredentials`,

Then we can catch the `WrongPasswordException` and get the `WrongPasswordException::user_id` from catched instance in `AccessControl::authenticate`. 

What do you think about this solution?

Reply:
Done!